torch/onnx/_internal/exporter.py

```diff
@@ -275,10 +275,12 @@ def enable_fake_mode():
     from torch.fx.experimental.symbolic_shapes import ShapeEnv
 
     # This overrides the internal `FakeTensorMode` instance created by `torch._dynamo.export`[1].
-    # Ideally we should keep them in sync to preserve the same default behavior
+    # It is a good idea to keep them in sync (constructor args) to maintain the same default behavior
     # [1] `torch/_dynamo/output_graph.py::InstructionTranslator::OutputGraph.__init__`
+    # Mixed fake/real tensors are only allowed when `torch.onnx.dynamo_export` is not called within `FakeTensorMode`
+    # This is needed because models can create new parameters during `forward(self, *args, **kwargs)` run
     fake_mode = fake_tensor.FakeTensorMode(
-        allow_non_fake_inputs=True,  # https://github.com/pytorch/pytorch/issues/105077
+        allow_non_fake_inputs=not torch._guards.detect_fake_mode(),
```

Comment:
you shouldn't need this anymore

Reply:
`allow_non_fake_inputs=True` still produces

```
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1467, in validate
    raise Exception(
Exception: Please convert all Tensors to FakeTensors first or instantiate FakeTensorMode with 'allow_non_fake_inputs'. Found in aten.embedding.default(Parameter(FakeTensor(..., size=(1024, 2), requires_grad=True)), tensor([...], size=(1, 3)))` 
```
in some tests (e.g. `test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxFakeTensorWithOnnxRuntime_op_level_debug_False_dynamic_shapes_False_load_checkpoint_during_init_True_export_within_fake_mode_False::test_large_scale_exporter_with_tiny_gpt2`)

Reply:
🤔 

Are you sure the exported model actually works if you run it under a fake mode with your inputs

Reply:
Yes, we make numerical validation between ONNX Runtime and PyTorch Eager run.

The failing tests are the ones in which the model/input load are done within fake tensor mode, but `torch._dynamo.export` is not called within the fake mode

Reply:
That is a repro:

```
    def test_capture_symbolic_tracing_outside_fake_mode(self):
        from torch._dynamo.output_graph import config
        from torch._subclasses import fake_tensor
        from torch.fx.experimental.symbolic_shapes import ShapeEnv

        class Model(torch.nn.Module):
            def __init__(self) -> None:
                super().__init__()
                self.linear = torch.nn.Linear(2, 2)
                self.linear2 = torch.nn.Linear(2, 2)

            def forward(self, x):
                out = self.linear(x)
                out = self.linear2(out)
                return out

        # User-instantiated FakeTensorMode
        fake_mode = fake_tensor.FakeTensorMode(
            allow_non_fake_inputs=False,
            allow_fallback_kernels=True,
            shape_env=ShapeEnv(
                allow_scalar_outputs=config.capture_scalar_outputs,
                allow_dynamic_output_shape_ops=config.capture_dynamic_output_shape_ops,
                frame_id=0,
            ),
        )
        # Fakefy input+model before exporting it
        with fake_mode:
            x = torch.rand(5, 2, 2)
            model = Model()

        # Export the model with fake inputs and parameters
        for aten_graph in [True, False]:
            graph_module, _ = torch._dynamo.export(model, x, aten_graph=aten_graph)
            self.assertTrue(
                isinstance(graph_module, torch.fx.GraphModule),
                msg="test_capture_symbolic_tracing_within_fake_mode_aten_graph_"
                + str(aten_graph),
            )
```
which raises
```
___________________________________________________________ ExportTests.test_capture_symbolic_tracing_within_fake_mode ____________________________________________________________
Traceback (most recent call last):
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/opt/pytorch/torch/testing/_internal/common_utils.py", line 2356, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/test/dynamo/test_export.py", line 3100, in test_capture_symbolic_tracing_within_fake_mode
    graph_module, _ = torch._dynamo.export(model, x, aten_graph=aten_graph)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1044, in export
    graph = make_fx(
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 809, in wrapped
    t = dispatch_trace(wrap_key(func, args, fx_tracer, pre_dispatch), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/opt/pytorch/torch/_compile.py", line 24, in inner
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 307, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 468, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 307, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 817, in trace
    (self.create_arg(fn(*args)),),
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 485, in wrapped
    out = f(*tensors)
  File "<string>", line 1, in <lambda>
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1040, in graph_with_interpreter
    return torch.fx.Interpreter(graph).run(*args)
  File "/opt/pytorch/torch/fx/interpreter.py", line 138, in run
    self.env[node] = self.run_node(node)
  File "/opt/pytorch/torch/fx/interpreter.py", line 195, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/opt/pytorch/torch/fx/interpreter.py", line 312, in call_module
    return submod(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 795, in module_call_wrapper
    return self.call_module(mod, forward, args, kwargs)
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 425, in call_module
    return forward(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 788, in forward
    return _orig_module_call(mod, *args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1522, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1531, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
  File "/opt/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1054, in __torch_dispatch__
    return func(*args, **kwargs)
  File "/opt/pytorch/torch/_ops.py", line 437, in __call__
    return self._op(*args, **kwargs or {})
  File "/opt/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1174, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1285, in dispatch
    ) = self.validate_and_convert_non_fake_tensors(func, converter, args, kwargs)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1479, in validate_and_convert_non_fake_tensors
    args, kwargs = tree_map_only(
  File "/opt/pytorch/torch/utils/_pytree.py", line 397, in tree_map_only
    return tree_map(map_only(ty)(fn), pytree)
  File "/opt/pytorch/torch/utils/_pytree.py", line 327, in tree_map
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/opt/pytorch/torch/utils/_pytree.py", line 327, in <listcomp>
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/opt/pytorch/torch/utils/_pytree.py", line 378, in inner
    return f(x)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1474, in validate
    assert x.fake_mode is self, "Mixing fake modes NYI"
AssertionError: Mixing fake modes NYI

While executing %l__self___linear : [num_users=1] = call_module[target=L__self___linear](args = (%l_x_,), kwargs = {})
Original traceback:
  File "/opt/pytorch/test/dynamo/test_export.py", line 3079, in forward
    out = self.linear(x)


To execute this test, run the following from the base repo dir:
     python test/dynamo/test_export.py -k test_capture_symbolic_tracing_within_fake_mode

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

Reply:
Reverting https://github.com/pytorch/pytorch/pull/105477/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911L937 fixes the last repro above (`test_capture_symbolic_tracing_outside_fake_mode`) with `allow_non_fake_inputs=False`, but `test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxFakeTensorWithOnnxRuntime_op_level_debug_False_dynamic_shapes_False_load_checkpoint_during_init_True_export_within_fake_mode_False::test_large_scale_exporter_with_tiny_gpt2` still fails

Reply:
What happens if you have the fake mode enabled when you export?